File: src/client/java/io/github/romeoahmed/cursedoath/client/render/domain/DomainSphere.java
package io.github.romeoahmed.cursedoath.client.render.domain;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.util.ARGB;
import net.minecraft.world.phys.Vec3;

final class DomainSphere {
    private static final int LONGITUDES = 64;
    private static final int LATITUDES = 32;
    private static final double VIEW_LONGITUDE = 0.555;
    private static final Vec3[][] POINTS = new Vec3[LATITUDES + 1][LONGITUDES + 1];

    static {
        for (int latitude = 0; latitude <= LATITUDES; latitude++) {
            double pitch = -Math.PI / 2 + latitude * Math.PI / LATITUDES;
            for (int longitude = 0; longitude <= LONGITUDES; longitude++) {
                double yaw = ((double) longitude / LONGITUDES - VIEW_LONGITUDE) * 2 * Math.PI;
                POINTS[latitude][longitude] =
                        new Vec3(Math.sin(yaw) * Math.cos(pitch), Math.sin(pitch), Math.cos(yaw) * Math.cos(pitch));
            }
        }
    }

    private DomainSphere() {}

    static void draw(PoseStack.Pose pose, VertexConsumer vertices, double radius, float age, boolean interior) {
        int color = interior ? ARGB.gray(VoidOpening.reveal(age)) : 0xFF080911;
        draw(pose, vertices, radius, interior, color);
    }

    static void backdrop(PoseStack.Pose pose, VertexConsumer vertices, double radius, int color) {
        draw(pose, vertices, radius, false, color);
    }

    private static void draw(PoseStack.Pose pose, VertexConsumer vertices, double radius, boolean interior, int color) {
        for (int latitude = 0; latitude < LATITUDES; latitude++)
            for (int longitude = 0; longitude < LONGITUDES; longitude++) {
                vertex(pose, vertices, radius, interior, color, latitude, longitude);
                vertex(pose, vertices, radius, interior, color, latitude + 1, longitude);
                vertex(pose, vertices, radius, interior, color, latitude + 1, longitude + 1);
                vertex(pose, vertices, radius, interior, color, latitude, longitude + 1);
            }
    }

    private static void vertex(
            PoseStack.Pose pose,
            VertexConsumer vertices,
            double radius,
            boolean interior,
            int color,
            int latitude,
            int longitude) {
        var point = POINTS[latitude][longitude];
        var vertex = vertices.addVertex(
                pose, (float) (point.x * radius), (float) (point.y * radius), (float) (point.z * radius));
        if (interior) {
            int shade = longitude == 0 || longitude == LONGITUDES ? 0xFF000000 : color;
            vertex.setUv((float) longitude / LONGITUDES, 1f - (float) latitude / LATITUDES)
                    .setColor(shade);
        } else vertex.setColor(color);
    }
}

File: src/client/kotlin/io/github/romeoahmed/cursedoath/client/render/domain/DomainSphere.kt
package io.github.romeoahmed.cursedoath.client.render.domain

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import net.minecraft.world.phys.Vec3
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

internal object DomainSphere {
    private const val LONGITUDES = 64
    private const val LATITUDES = 32
    private const val CHANNEL = 255
    private const val REVEAL_TICKS = 24f
    private const val VIEW_LONGITUDE = 0.555
    private val points =
        Array(LATITUDES + 1) { latitude ->
            val pitch = -PI / 2 + latitude * PI / LATITUDES
            Array(LONGITUDES + 1) { longitude ->
                val yaw = (longitude.toDouble() / LONGITUDES - VIEW_LONGITUDE) * 2 * PI
                Vec3(sin(yaw) * cos(pitch), sin(pitch), cos(yaw) * cos(pitch))
            }
        }

    fun draw(
        pose: PoseStack.Pose,
        vertices: VertexConsumer,
        radius: Double,
        age: Float,
        interior: Boolean,
    ) {
        val alpha = ((age / REVEAL_TICKS).coerceIn(0f, 1f) * CHANNEL).toInt()

        fun vertex(
            latitude: Int,
            longitude: Int,
        ) {
            val point = points[latitude][longitude]
            val vertex =
                vertices.addVertex(
                    pose,
                    (point.x * radius).toFloat(),
                    (point.y * radius).toFloat(),
                    (point.z * radius).toFloat(),
                )
            if (interior) {
                vertex
                    .setUv(longitude.toFloat() / LONGITUDES, 1f - latitude.toFloat() / LATITUDES)
                    .setColor(CHANNEL, CHANNEL, CHANNEL, alpha)
            } else {
                vertex.setColor(0xFF080911.toInt())
            }
        }
        for (latitude in 0..<LATITUDES) {
            for (longitude in 0..<LONGITUDES) {
                vertex(latitude, longitude)
                vertex(latitude + 1, longitude)
                vertex(latitude + 1, longitude + 1)
                vertex(latitude, longitude + 1)
            }
        }
    }
}
